File: src/SOS.Bidnez/Character.cs
using Microsoft.FSharp.Core;
using SOS.Persistence;
using System;

namespace SOS.Bidnez
{
    internal class Character : ICharacter
    {
        private WorldData _worldData;
        private int _id;
        internal Character(WorldData worldData, int id)
        {
            _worldData = worldData;
            _id = id;
        }

        public string Name => _worldData.Characters[_id].Name;

        public double X => _worldData.Characters[_id].X;

        public double Y => _worldData.Characters[_id].Y;

        public int Id => _id;

        internal static ICharacter? FromId(WorldData worldData, FSharpOption<int> id)
        {
            return (id is null) ? (null) : new Character(worldData, id.Value);
        }
    }
}

File: src/SOS.Business/Character.vb
Public Class Character
    Implements ICharacter
    Private ReadOnly _id As Integer
    Private ReadOnly _worldData As WorldData
    Sub New(worldData As WorldData, id As Integer)
        _worldData = worldData
        _id = id
    End Sub

    Public ReadOnly Property Id As Integer Implements ICharacter.Id
        Get
            Return _id
        End Get
    End Property

    Public ReadOnly Property Name As String Implements ICharacter.Name
        Get
            Return _worldData.Characters(Id).Name
        End Get
    End Property

    Public ReadOnly Property X As Double Implements ICharacter.X
        Get
            Return _worldData.Characters(Id).X
        End Get
    End Property

    Public ReadOnly Property Y As Double Implements ICharacter.Y
        Get
            Return _worldData.Characters(Id).Y
        End Get
    End Property

    Public Property Heading As Double Implements ICharacter.Heading
        Get
            Return _worldData.Characters(Id).Heading
        End Get
        Set(value As Double)
            _worldData.Characters(Id).Heading = Math.Min(360.0, Math.Max(0.0, value))
        End Set
    End Property

    Public Property Speed As Double Implements ICharacter.Speed
        Get
            Return _worldData.Characters(Id).Speed
        End Get
        Set(value As Double)
            _worldData.Characters(Id).Speed = Math.Min(1.0, Math.Max(0.0, value))
        End Set
    End Property
    Friend Shared Function FromId(worldData As WorldData, id As Integer?) As ICharacter
        Return If(id.HasValue, New Character(worldData, id.Value), Nothing)
    End Function
End Class
